modules/priceFloors.js

```diff
@@ -0,0 +1,645 @@
+    return accum;
+  }, []);
+}
+
+/**
+ * @summary get's the first matching floor based on context provided.
+ * Generates all possible rule matches and picks the first matching one.
+ */
+export function getFirstMatchingFloor(floorData, bidObject, responseObject = {}) {
+  let fieldValues = enumeratePossibleFieldValues(utils.deepAccess(floorData, 'schema.fields') || [], bidObject, responseObject);
+  if (!fieldValues.length) return { matchingFloor: floorData.default };
+
+  // look to see iof a request for this context was made already
+  let matchingInput = fieldValues.map(field => field[0]).join('-');
+  // if we already have gotten the matching rule from this matching input then use it! No need to look again
+  let previousMatch = utils.deepAccess(floorData, `matchingInputs.${matchingInput}`);
+  if (previousMatch) {
+    return previousMatch;
+  }
+  let allPossibleMatches = generatePossibleEnumerations(fieldValues, utils.deepAccess(floorData, 'schema.delimiter') || '|');
+  let matchingRule = find(allPossibleMatches, hashValue => floorData.values.hasOwnProperty(hashValue));
+
+  let matchingData = {
+    matchingFloor: floorData.values[matchingRule] || floorData.default,
+    matchingData: allPossibleMatches[0], // the first possible match is an "exact" so contains all data relevant for anlaytics adapters
+    matchingRule
+  };
+  // save for later lookup if needed
+  utils.deepSetValue(floorData, `matchingInputs.${matchingInput}`, {...matchingData});
+  return matchingData;
+}
+
+/**
+ * @summary Generates all possible rule hash's based on input array of array's
+ * The generated list is of all possible key matches based on fields input
+ * The list is sorted by least amount of * in rule to most with left most fields taking precedence
+ */
+function generatePossibleEnumerations(arrayOfFields, delimiter) {
+  return arrayOfFields.reduce((accum, currentVal) => {
+    let ret = [];
+    accum.map(obj => {
+      currentVal.map(obj1 => {
+        ret.push(obj + delimiter + obj1)
+      });
+    });
+    return ret;
+  }).sort((left, right) => left.split('*').length - right.split('*').length);
+}
+
+/**
+ * @summary If a the input bidder has a registered cpmadjustment it returns the input CPM after being adjusted
+ */
+export function getBiddersCpmAdjustment(bidderName, inputCpm) {
+  const adjustmentFunction = utils.deepAccess(getGlobal(), `bidderSettings.${bidderName}.bidCpmAdjustment`);
+  if (adjustmentFunction) {
+    return parseFloat(adjustmentFunction(inputCpm));
+  }
+  return parseFloat(inputCpm);
+}
+
+/**
+ * @summary This function takes the original floor and the adjusted floor in order to determine the bidders actual floor
+ * With js rounding errors with decimal division we utilize similar method as shown in cpmBucketManager.js
+ */
+export function calculateAdjustedFloor(oldFloor, newFloor) {
+  const pow = Math.pow(10, 10);
+  return (oldFloor * pow / newFloor * oldFloor) / pow;
```

Comment:
I'm not sure this updated formula is correct.  Wouldn't we still be dividing by decimals via the `newFloor` var here?

I was thinking we'd have to multiply each value of the original formula by the `pow` and then divide the result by `pow` again to reduce it to normal numbers.  Something perhaps like `((oldFloor * pow) / (newFloor * pow) * (oldFloor * pow)) / pow`.

Reply:
Ok yeah I updated. Appreciate the insight, I have not been exposed to this issue before so glad to learn something!